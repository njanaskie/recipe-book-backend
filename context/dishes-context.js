import React from 'react';

const DishesContext = React.createContext()

export { DishesContext as default }