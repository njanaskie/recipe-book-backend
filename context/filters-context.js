import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../src/reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }