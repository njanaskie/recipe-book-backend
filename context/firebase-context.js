import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../src/firebase/firebase';
import { history } from '../src/routers/AppRouter';
import authReducer from '../src/reducers/auth'
import LoadingPage from '../src/components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    // const authDispatch = useDispatch()
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/dashboard')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }