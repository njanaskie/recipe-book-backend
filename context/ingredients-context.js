import React from 'react';

const IngredientsContext = React.createContext()

export { IngredientsContext as default }