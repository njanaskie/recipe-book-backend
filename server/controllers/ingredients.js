const ingredientsRouter = require('express').Router();
const Ingredient = require('../models/ingredient')


getIngredients = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        const ingredients = await Ingredient.find({}).sort( { name: 1 } )
        return res.json(ingredients.map((ingredient) => ingredient.toJSON() ));
    }
    return res.status(403).send('Not authorized get')
}
// ingredientsRouter.get('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//         // const ingredient = await Ingredient.find({})
//         // console.log('authenticated!', auth);
//         return res.send('Hi, from within the ingredients router GET'); 
//     // }
//     // console.log(req.currentUser)
//     // return res.status(403).send('Not authorized get')
// });

createIngredient = (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        // console.log('authenticated!', auth);
        const ingredient = new Ingredient(req.body)
        const savedIngredient = ingredient.save()

        return res.status(201).json(savedIngredient);
    }
    return res.status(403).send('Not authorized post')

}
// ingredientsRouter.post('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//     //     console.log('authenticated!', auth);
//         const ingredient = new Ingredient(req.body)
//         const savedIngredient = ingredient.save()

//         return res.status(201).json(savedIngredient);
//     // }
//     // return res.status(403).send('Not authorized post')
// });

removeIngredient = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {

        await Ingredient.findOneAndDelete({ _id: req.params.id }, (err, ingredient) => {
            if (err) {
                return res.status(400).json({ success: false, error: err })
            }

            if (!ingredient) {
                return res
                    .status(404)
                    .json({ success: false, error: `Ingredient not found` })
            }

            return res.status(200).json({ success: true, data: ingredient })
        }).catch(err => console.log(err))

    }
    return res.status(403).send('Not authorized post')

}

module.exports = {
    getIngredients,
    createIngredient,
    removeIngredient
};