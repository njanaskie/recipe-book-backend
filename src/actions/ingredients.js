import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_INGREDIENT
export const addIngredient = (ingredient) => ({
    type: 'ADD_INGREDIENT',
    ingredient
});

export const startAddIngredient = (ingredientData = {}) => {
    return (dispatch) => {
        const {
            name = '',
            category = '',
            price = ''
        } = ingredientData;
        const ingredient = { name, category, price };

        return database.ref('ingredients').push(ingredient).then((ref) => {
            dispatch(addIngredient({
                id: ref.key,
                ...ingredient
            }));
        });
    };
};

// REMOVE_INGREDIENT
export const removeIngredient = ({ id } = {}) => ({
    type: 'REMOVE_INGREDIENT',
    id
});

export const startRemoveIngredient = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).remove().then(() => {
            dispatch(removeIngredient({ id }))
        });
    };
};

// EDIT_INGREDIENT
export const editIngredient = (id, updates) => ({
    type: 'EDIT_INGREDIENT',
    id,
    updates
});

export const startEditIngredient = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).update(updates).then(() => {
            dispatch(editIngredient(id, updates))
        });
    }
}

// SET_INGREDIENTS
export const setIngredients = (ingredients) => ({
    type: 'SET_INGREDIENTS',
    ingredients
});

export const startSetIngredients = () => {
    return (dispatch, getState) => {
        return database.ref('ingredients')
        .once('value')
        .then((snapshot) => {
            const ingredients = [];

            snapshot.forEach((childSnapshot) => {
                ingredients.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setIngredients(ingredients))
        });
    };
};