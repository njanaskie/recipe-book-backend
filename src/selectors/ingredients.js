
export default (ingredients) => {
    return ingredients.filter(ingredient => ingredient.id !== undefined
    )
    .sort((a, b) => {
        if (a.category === b.category) {
            return b.category - a.category
        }
         return a.category > b.category ? 1 : -1
    })
}