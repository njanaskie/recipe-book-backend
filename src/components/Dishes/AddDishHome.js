import React, { useContext, useEffect } from 'react'
import { useHistory } from 'react-router-dom'
import { useDishesContext } from '../../../context/dishes-context'
import DishForm from '../Dishes/DishForm'
import useDishes from '../../hooks/useDishes'
import database from '../../firebase/firebase'

export const AddDishHome = (props) => {
    const { dishDispatch } = useDishesContext()
    const history = useHistory()

    const onSubmit = (dish) => {
        database.collection('dishes').add(dish).then((ref) => {
            dishDispatch({ type: 'ADD_DISH', dish: {id: ref.key, ...dish} })
            history.push('/dishes')
        })
    }

    return (
        <div>
            <DishForm
                onSubmit={onSubmit}
            />
        </div>
    )
}

export default AddDishHome