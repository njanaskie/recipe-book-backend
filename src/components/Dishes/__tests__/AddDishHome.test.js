import React from 'react'
import { shallow, mount } from 'enzyme'
import { Router } from 'react-router-dom';
import * as router from 'react-router';
import ingredients from '../../../tests/fixtures/ingredients'
import { AddDishHome } from '../AddDishHome'
import DishesContext from '../../../../context/dishes-context'
import * as AllDishesContext from '../../../../context/dishes-context'
import IngredientsContext from '../../../../context/ingredients-context'

let mockDishDispatch, mockHistory;

// mockHistory = jest.mock('react-router-dom', () => ({
//     ...jest.requireActual('react-router-dom'),
//     useHistory: () => ({
//       push: jest.fn()
//     })
//   }));
  
beforeEach(() => {
    mockDishDispatch = jest.fn()
    // mockHistory = { push: jest.fn() }
    mockHistory = { push: jest.fn(), location: {}, listen: jest.fn() };
    // mockHistory = jest.mock('react-router-dom', () => ({
    //     useHistory: () => ({
    //       push: jest.fn(),
    //       history: { listen: jest.fn() },
    //       location: {},
    //       listen: jest.fn()
    //     })
    //   }));

})

afterEach(() => {
    jest.clearAllMocks();
})


test('should render AddDishHome correctly',() => {
    jest.spyOn(AllDishesContext, 'useDishesContext').mockImplementation(() => mockDishDispatch)
    jest.spyOn(router, 'useHistory').mockImplementation(() => mockHistory);
    const wrapper = shallow(
        <Router history={mockHistory}>
            <DishesContext.Provider value={{ dishDispatch: mockDishDispatch }}>
                <IngredientsContext.Provider value={{ ingredients }}>
                    <AddDishHome />
                </IngredientsContext.Provider>
            </DishesContext.Provider>
        </Router>
    )
    // .find('AddDishHome').dive()
    expect(wrapper).toMatchSnapshot()
})

// test('should handle onSubmit',() => {
//     const wrapper = mount(
//         <Router history={mockHistory}>
//             <DishesContext.Provider value={{ dishDispatch: mockDishDispatch }}>
//                 <IngredientsContext.Provider value={{ ingredients }}>
//                     <AddDishHome />
//                 </IngredientsContext.Provider>
//             </DishesContext.Provider>
//         </Router>
//     )
//     wrapper.find('DishForm').prop('onSubmit')(ingredients[1])
//     expect(mockHistory.push).toHaveBeenLastCalledWith('/dishes')
//     // expect(startAddExpense).toHaveBeenLastCalledWith(ingredients[1])
// })