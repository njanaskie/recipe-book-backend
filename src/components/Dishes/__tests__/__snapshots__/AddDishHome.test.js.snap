// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddDishHome correctly 1`] = `
<ContextProvider
  value={
    Object {
      "history": Object {
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "location": Object {},
        "push": [MockFunction],
      },
      "location": Object {},
      "match": Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      },
      "staticContext": undefined,
    }
  }
>
  <ContextProvider
    value={
      Object {
        "dishDispatch": [MockFunction],
      }
    }
  >
    <ContextProvider
      value={
        Object {
          "ingredients": Array [
            Object {
              "category": "Spices",
              "id": "1",
              "name": "Salt",
              "price": 1,
            },
            Object {
              "category": "Vegetables",
              "id": "2",
              "name": "Lettuce",
              "price": 1,
            },
            Object {
              "category": "Dairy",
              "id": "3",
              "name": "Yoghurt",
              "price": 2,
            },
          ],
        }
      }
    >
      <AddDishHome />
    </ContextProvider>
  </ContextProvider>
</ContextProvider>
`;
